### ISSUE TITLE
BioPerl build process migrated to using Dist::Zilla

### ISSUE THREAD

**Issue by hlapp** [bioperl/bioperl-docker#5]
BioPerl migrated its build and installation process to using Dist::Zilla. As a result, installing from source  by simply using `cpanm` fails now.

See bioperl/bioperl-live#289 for details and how to possibly fix this.

**Comment by carandraug** [comment]
Also important is the fact that many of the modules in the BioPerl distribution have been moved into a different distribution. I'm not sure what is the aim of this project but it may make sense to install all distributions.

**Comment by hlapp** [comment]
> I'm not sure what is the aim of this project

A reproducible, documented, versioned, and ready to use BioPerl installation.

> but it may make sense to install all distributions.

Yes, that would be most consistent with its original goals and content.

**Comment by carandraug** [comment]
On Sat, 29 Sep 2018 at 20:26, Hilmar Lapp <notifications@github.com> wrote:
>
>> I'm not sure what is the aim of this project
>
> A reproducible, documented, versioned, and ready to use BioPerl installation.

Sounds fancy. There's a BioPerl (umbrella) project/organization,
there's a `BioPerl` module, and there's a BioPerl module distribution.

>> but it may make sense to install all distributions.
>
> Yes, that would be most consistent with its original goals and content.

All distributions include BioPerl-Run?

There's been movement in many directions. Thing moving from BioPerl to
BioPerl-Run, things from BioPerl-Run into BioPerl, and things from
both getting together to make new distributions.

**Comment by yzhernand** [comment]
Hi Hilmar, I decided to give a shot at helping out with this. I forked the repository and my solution seems to work (https://github.com/yzhernand/bioperl-docker). I'm using multistage builds to keep all the modules that are only needed for building bioperl out of the final image.

I think I need some help testing out if the final image works, so I can make a pull request. How do you test if it works as expected? I can test that the perl within the image finds the bioperl modules alright, but that's about it.

**Comment by hlapp** [comment]
@yzhernand that's an interesting idea to use a multistage build Dockerfile. I wasn't aware that is now possible.

Previously, the tests performed was the test suite, which you're disabling here (through `-n`). But that's easy to change. And indeed there are test failures right now on master.

So please go ahead and make a pull request.

**Comment by yzhernand** [comment]
Oh yes, I disabled tests because it takes a very long time. And I figured they were safe enough to skip since this is building from bleeding edge, after all. At least for the BioPerl code.

I actually meant how do I test that the whole image works as intend. I basically taught myself docker just for this so I have no idea for to use the image I made! I just know how to check what is in it.

I'll make the pull request.

**Comment by hlapp** [comment]
One of the main use-cases for the image is for someone who doesn't have BioPerl installed to be able to run a piece of Perl software that depends on BioPerl. Typically this would happen by binding a volume to the container that contains the code to be run and the data on which it is to be run, and one would use `docker run` to run that command. Or one can shell into the container. Whatever is needed.

Hence, to test the container, one would run the BioPerl test suite as part of the cpanm installation, or by shelling into the container and running them then. I did the latter, which shows a couple of test failures, such as bioperl/bioperl-live#292 and bioperl/bioperl-live#293.

**Comment by hlapp** [comment]
With the results of #6 and subsequent creation of a bioperl/bioperl-builddeps container we now have a functioning build of BioPerl from source that uses Dist::Zilla.

There are remaining issues, specifically test failures (see #7) and remaining dependency on BioPerl v1.7.2 from CPAN. The former is already getting addressed, and the latter is a part of completing the migration to Dist::Zilla, which was expected to take some time.

In conclusion, I think we can close this one out, as the remaining issues above go beyond the initial scope here of just getting the build corrected for Dist::Zilla.
